consumer.go

```diff
@@ -171,31 +171,39 @@ func (c *consumer) refBrokerConsumer(broker *Broker) *brokerConsumer {
 			newSubscriptions: make(chan []*partitionConsumer),
 			wait:             make(chan none),
 			subscriptions:    make(map[*partitionConsumer]none),
-			refs:             1,
+			refs:             0,
 		}
 		go withRecover(brokerWorker.subscriptionManager)
 		go withRecover(brokerWorker.subscriptionConsumer)
 		c.brokerConsumers[broker] = brokerWorker
-	} else {
-		brokerWorker.refs++
 	}
 
+	brokerWorker.refs++
+
 	return brokerWorker
 }
 
-func (c *consumer) unrefBrokerConsumer(broker *Broker) {
+func (c *consumer) unrefBrokerConsumer(brokerWorker *brokerConsumer) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	brokerWorker := c.brokerConsumers[broker]
 	brokerWorker.refs--
 
 	if brokerWorker.refs == 0 {
 		close(brokerWorker.input)
-		delete(c.brokerConsumers, broker)
+		if c.brokerConsumers[brokerWorker.broker] == brokerWorker {
+			delete(c.brokerConsumers, brokerWorker.broker)
+		}
```

Comment:
It triggers when all PartitionConsumers move off of the brokerConsumer "safely", either via being closed or via a `NotLeaderForPartition` response. When the broker blows up we abort, and the delete happens immediately to prevent new partitions from taking a reference.